FEATURE: Fit SideTitle inside axis bounding box (prevent title overflow on edge)

dhiyaaulauliyaa opened the pull request:
Hi All!

I'm adding a feature to add an option for SideTitle to be fitted inside its corresponding axis bounding box.
This will fix #603. Below is the illustration:
- Before Using FitInside (SideTitle is overflowed): 
![Screenshot_1675854250](https://user-images.githubusercontent.com/20693268/217512793-d1c54030-a5a9-4d63-afed-86c93add8794.png)
- After Using FitInside (SideTitle is translated, hence not overflowed):
![Screenshot_1675854231](https://user-images.githubusercontent.com/20693268/217512828-2aeaf7a8-ccd7-409e-980a-da516e5ebd32.png)

What is added:
- FitInside data class (SideTitleFitInsideData) in lib/src/chart/base/axis_chart/axis_chart_data.dart
- FitInside parameter on SideTitleWidget
- Test for data class (SideTitleFitInsideData)
- Sample usage in example/lib/presentation/samples/line/line_chart_sample3.dart

imaNNeo commented:
Thank you for this great feature!
First of all, please check the below bug:
When I hover my mouse, it hides the titles. 
https://user-images.githubusercontent.com/7009300/217529221-63d57390-f75c-4799-96e1-7269a8d61c93.mov

It seems it is related to this code:
```dart
ValueListenableBuilder(
  valueListenable: _childSize,
  builder: (_, double? size, Widget? child) {
    return Transform.translate(
      offset: _getOffset(),
      child: Opacity(
        opacity: fitInside.enabled && size == null ? 0 : 1,
        child: child,
      ),
    );
  },
)
```

Also, there are some failing CI workflows. Please check and fix those issues.
Don't forget to update the documentation. Also, you should write a changelog for it.
So it would be great if you read our [contributing guidelines](https://github.com/imaNNeo/fl_chart/blob/master/CONTRIBUTING.md).

dhiyaaulauliyaa commented:
> Thank you for this great feature! First of all, please check the below bug: When I hover my mouse, it hides the titles.
> 
>  hover-bug.mov 
> It seems it is related to this code:
> 
> ```dart
> ValueListenableBuilder(
>   valueListenable: _childSize,
>   builder: (_, double? size, Widget? child) {
>     return Transform.translate(
>       offset: _getOffset(),
>       child: Opacity(
>         opacity: fitInside.enabled && size == null ? 0 : 1,
>         child: child,
>       ),
>     );
>   },
> )
> ```
> 
> Also, there are some failing CI workflows. Please check and fix those issues. Don't forget to update the documentation. Also, you should write a changelog for it. So it would be great if you read our [contributing guidelines](https://github.com/imaNNeo/fl_chart/blob/master/CONTRIBUTING.md).

Thanks for the response.

The bug has been fixed. Now the title won't blink when the chart is on hover.
Also, I have added the widget test and increased the coverage of my added changes.
For the docs, I have updated the documentation and added a changelog about this added feature.

imaNNeo commented on lib/src/chart/base/axis_chart/axis_chart_helper.dart:
```diff
@@ -51,4 +53,52 @@ class AxisChartHelper {
+
+    // Find title alignment along its axis
+    final axisMid = parentAxisSize / 2;
+    final mainAxisAligment = (axisPosition - axisMid).isNegative
```
There is a typo issue in `mainAxisAligment`. Please fix it.

dhiyaaulauliyaa replied:
I have fix the typo in the latest commit

imaNNeo commented:
Hi, we are ready to merge after fixing the typo. 

Thanks!

imaNNeo commented:
Thank you very much.

HornyPony commented:
Hi! How can I usee this option?
